pkg/agent/flags/flag_controller.go

```diff
@@ -0,0 +1,189 @@
+	controlRequestOverride FlagValueOverride
+	observers              map[types.FlagsChangeObserver][]keys.FlagKey
+}
+
+func NewFlagController(logger log.Logger, agentFlagsStore types.KVStore, opts ...Option) *FlagController {
+	fc := &FlagController{
+		logger:          logger,
+		cmdLineOpts:     &launcher.Options{},
+		agentFlagsStore: agentFlagsStore,
+		observers:       make(map[types.FlagsChangeObserver][]keys.FlagKey),
+	}
+
+	for _, opt := range opts {
+		opt(fc)
+	}
+
+	return fc
+}
+
+// getControlServerValue looks for a control-server-provided value for the key and returns it.
+// If a control server value is not found, nil is returned.
+func (fc *FlagController) getControlServerValue(key keys.FlagKey) []byte {
+	value, err := fc.agentFlagsStore.Get([]byte(key))
+	if err != nil {
+		level.Debug(fc.logger).Log("msg", "failed to get control server key", "key", key, "err", err)
+		return nil
+	}
+
+	return value
+}
+
+// set stores the value for a FlagKey in the agent flags store. Typically, this is used by the control
+// server to provide a control-server-provided value.
+func (fc *FlagController) set(key keys.FlagKey, value []byte) error {
+	err := fc.agentFlagsStore.Set([]byte(key), value)
+	if err != nil {
+		level.Debug(fc.logger).Log("msg", "failed to set control server key", "key", key, "err", err)
+		return err
+	}
+
+	fc.notifyObservers(key)
+
+	return nil
+}
+
+// Update bulk replaces agent flags and stores them.
+// Observers will be notified of changed flags and deleted flags.
+func (fc *FlagController) Update(kvPairs map[string]string) ([]string, error) {
+	// Attempt to bulk replace the store with the key-values
+	deletedKeys, err := fc.agentFlagsStore.Update(kvPairs)
+
+	// Extract just the keys from the key-value pairs
+	updatedKeys := make([]string, len(kvPairs))
+	i := 0
+	for k := range kvPairs {
+		updatedKeys[i] = k
+		i++
+	}
+
+	// Changed keys is the union of updated keys and deleted keys
+	changedKeys := append(updatedKeys, deletedKeys...)
+
+	// Now observers can be notified these keys have possibly changed
+	fc.notifyObservers(keys.ToFlagKeys(changedKeys)...)
+
+	return changedKeys, err
+}
+
+func (fc *FlagController) RegisterChangeObserver(observer types.FlagsChangeObserver, flagKeys ...keys.FlagKey) {
+	fc.observers[observer] = append(fc.observers[observer], flagKeys...)
+}
+
+// notifyObservers informs all observers of the keys that they have changed.
+func (fc *FlagController) notifyObservers(flagKeys ...keys.FlagKey) {
+	for observer, observedKeys := range fc.observers {
+		changedKeys := keys.Intersection(observedKeys, flagKeys)
+
+		if len(changedKeys) > 0 {
+			observer.FlagsChanged(changedKeys...)
+		}
+	}
+}
+
+func (fc *FlagController) SetDesktopEnabled(enabled bool) error {
+	return fc.set(keys.DesktopEnabled, boolToBytes(enabled))
+}
+func (fc *FlagController) DesktopEnabled() bool {
+	return NewBoolFlagValue(WithDefaultBool(false)).get(fc.getControlServerValue(keys.DesktopEnabled))
```

Comment:
I think this is fine, but I also wonder about the underlying performance. 

`NewBoolFlagValue(WithDefaultBool(false)).get` looks like it does a bunch of allocation and option parsing. Vs `BoolFlagValue(fc.getControlServerValue(keys.DesktopEnabled, false)`. But I also think I'm probably over thinking this, since we always have to do a bunch of conversations and storage reads anyhow. 

So probably optimizing for developer UX is good. Let's try!